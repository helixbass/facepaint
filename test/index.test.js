/* eslint-disable no-sparse-arrays */
import React from 'react'
import renderer from 'react-test-renderer'
import {css, sheet, flush, cx} from 'emotion'

import facepaint from '../src/index'

const mq = facepaint(
  {
    first: '@media(min-width: 420px)',
    second: '@media(min-width: 920px)',
    third: '@media(min-width: 1120px)',
    fourth: '@media(min-width: 11200px)',
  },
  {default: 'zero', overlap: true}
)

const pseudo = facepaint({first: ':hover', second: ':active', third: ':focus'})

describe('facepaint', () => {
  afterEach(() => flush())
  test('basic', () => {
    const result = css(
      mq({
        color: {
          zero: 'red',
          first: 'green',
          second: 'blue',
          third: 'darkorchid',
        },
      })
    )
    const tree = renderer.create(<div css={result}>Basic</div>).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('holes', () => {
    const result = css(
      mq({color: {zero: 'red', second: 'blue', third: 'darkorchid'}})
    )
    const tree = renderer.create(<div css={result}>Basic</div>).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('undefined', () => {
    const result = css(
      mq({
        color: {
          zero: 'red',
          first: undefined,
          second: 'blue',
          third: 'darkorchid',
        },
      })
    )
    const tree = renderer.create(<div css={result}>Basic</div>).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('repeating', () => {
    const result = css(
      mq({
        color: {
          zero: 'red',
          first: 'blue',
          third: 'blue',
          fourth: 'darkorchid',
        },
      })
    )
    const tree = renderer.create(<div css={result}>Basic</div>).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('nested arrays', () => {
    const result = css(
      mq([[[[{color: {zero: 'red', first: 'blue', second: 'darkorchid'}}]]]])
    )
    const tree = renderer.create(<div css={result}>Basic</div>).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('multiple', () => {
    const result = css(
      mq({
        color: {
          zero: 'red',
          first: 'green',
          second: 'blue',
          third: 'darkorchid',
        },
        display: {
          zero: 'flex',
          first: 'block',
          second: 'inline-block',
          third: 'table',
        },
        fontSize: 12,
        alignItems: 'center',
      })
    )
    const tree = renderer.create(<div css={result}>multiple</div>).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('nested', () => {
    const result = css(
      mq({
        backgroundColor: 'hotpink',
        textAlign: 'center',
        width: {zero: '25%', first: '50%', second: '75%', third: '100%'},
        '& .foo': {
          color: {
            zero: 'red',
            first: 'green',
            second: 'blue',
            third: 'darkorchid',
          },
          '& img': {
            height: {zero: 10, first: 15, second: 20, third: 25},
          },
        },
      })
    )
    const tree = renderer
      .create(
        <div css={result}>
          <div className="foo">foo</div>
          function
        </div>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('pseudo', () => {
    const result = css(
      pseudo({
        backgroundColor: 'hotpink',
        textAlign: 'center',
        width: {default: '25%', first: '50%', second: '75%', third: '100%'},
        '& .foo': {
          color: {
            default: 'red',
            first: 'green',
            second: 'blue',
            third: 'darkorchid',
          },
          '& img': {
            height: {default: 10, first: 15, second: 20, third: 25},
          },
        },
      })
    )
    const tree = renderer
      .create(
        <div css={result}>
          <div className="foo">foo</div>
          function
        </div>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('array values with selectors', () => {
    const result = css(
      mq({
        '& .current-index': [
          {
            color: {zero: 'blue', first: 'red'},
          },
          {
            marginRight: 15,
            display: {zero: 'none', first: 'block'},
            letterSpacing: 3,
          },
        ],
      })
    )
    const tree = renderer
      .create(
        <div css={result}>
          <div className="foo">foo</div>
          function
        </div>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('boolean, null, and undefined values', () => {
    const result = css(
      mq(
        {color: 'blue'},
        1 === 2 && {color: 'green'},
        false,
        true,
        undefined,
        null,
        [
          {color: 'red'},
          1 === 2 && {color: 'green'},
          false,
          true,
          undefined,
          null,
        ]
      )
    )

    const tree = renderer
      .create(
        <div css={result}>
          <div className="foo">foo</div>
          function
        </div>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('composition', () => {
    const a = css(mq({background: {zero: 'green', first: 'blue'}}))
    const b = css(mq({background: 'orange'}))
    const c = css(mq({background: {zero: 'orange', first: 'orange'}}))
    const d = css(
      mq({
        background: {
          zero: 'orange',
          first: 'orange',
          second: 'orange',
          third: 'orange',
        },
      })
    )

    const tree = renderer.create(<div css={cx(a, b, c, d)} />).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('more composition', () => {
    const styles1 = css(mq({marginTop: {zero: 1, first: 2}}))
    const styles2 = css(mq({marginTop: {zero: 500, first: 500}}))
    const tree = renderer.create(<div css={cx(styles1, styles2)} />).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('complex overlapped', () => {
    const result = mq({
      color: {zero: 'red', first: 'red', second: 'blue', third: 'darkorchid'},
      background: {
        zero: 'red',
        first: 'green',
        second: 'green',
        third: 'darkorchid',
      },
    })
    expect(result).toMatchSnapshot()
    expect(Object.keys(result[0])).toMatchSnapshot()
  })
})
