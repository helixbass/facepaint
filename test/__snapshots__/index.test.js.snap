// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`facepaint array values with selectors 1`] = `
.emotion-0 .current-index {
  color: blue;
  margin-right: 15px;
  display: none;
  -webkit-letter-spacing: 3px;
  -moz-letter-spacing: 3px;
  -ms-letter-spacing: 3px;
  letter-spacing: 3px;
}

@media (min-width:420px) {
  .emotion-0 .current-index {
    color: red;
  }
}

@media (min-width:420px) {
  .emotion-0 .current-index {
    display: block;
  }
}

<div
  className="emotion-0"
>
  <div
    className="foo"
  >
    foo
  </div>
  function
</div>
`;

exports[`facepaint array values with selectors 2`] = `
".css-j9a8v3 .current-index {
  color: blue;
  margin-right: 15px;
  display: none;
  -webkit-letter-spacing: 3px;
  -moz-letter-spacing: 3px;
  -ms-letter-spacing: 3px;
  letter-spacing: 3px;
}

@media (min-width:420px) {
  .css-j9a8v3 .current-index {
    color: red;
  }
}

@media (min-width:420px) {
  .css-j9a8v3 .current-index {
    display: block;
  }
}"
`;

exports[`facepaint basic 1`] = `
.emotion-0 {
  color: red;
}

@media (min-width:420px) {
  .emotion-0 {
    color: green;
  }
}

@media (min-width:920px) {
  .emotion-0 {
    color: blue;
  }
}

@media (min-width:1120px) {
  .emotion-0 {
    color: darkorchid;
  }
}

<div
  className="emotion-0"
>
  Basic
</div>
`;

exports[`facepaint basic 2`] = `
".css-xee1ha {
  color: red;
}

@media (min-width:420px) {
  .css-xee1ha {
    color: green;
  }
}

@media (min-width:920px) {
  .css-xee1ha {
    color: blue;
  }
}

@media (min-width:1120px) {
  .css-xee1ha {
    color: darkorchid;
  }
}"
`;

exports[`facepaint boolean, null, and undefined values 1`] = `
.emotion-0 {
  color: blue;
  color: red;
}

<div
  className="emotion-0"
>
  <div
    className="foo"
  >
    foo
  </div>
  function
</div>
`;

exports[`facepaint boolean, null, and undefined values 2`] = `
".css-dhb7kq {
  color: blue;
  color: red;
}"
`;

exports[`facepaint complex overlapped 1`] = `
Array [
  Object {
    "@media(min-width: 1120px)": Object {
      "background": "darkorchid",
      "color": "darkorchid",
    },
    "@media(min-width: 420px)": Object {
      "background": "green",
    },
    "@media(min-width: 920px)": Object {
      "color": "blue",
    },
    "background": "red",
    "color": "red",
  },
]
`;

exports[`facepaint complex overlapped 2`] = `
Array [
  "color",
  "background",
  "@media(min-width: 420px)",
  "@media(min-width: 920px)",
  "@media(min-width: 1120px)",
]
`;

exports[`facepaint composition 1`] = `
.emotion-0 {
  background: green;
  background: orange;
  background: orange;
  background: orange;
}

@media (min-width:420px) {
  .emotion-0 {
    background: blue;
  }
}

<div
  className="emotion-0"
/>
`;

exports[`facepaint composition 2`] = `
".css-16gx40p {
  background: green;
}

@media (min-width:420px) {
  .css-16gx40p {
    background: blue;
  }
}

.css-1y2rzxt {
  background: orange;
}

.css-1plng0y {
  background: green;
  background: orange;
  background: orange;
  background: orange;
}

@media (min-width:420px) {
  .css-1plng0y {
    background: blue;
  }
}"
`;

exports[`facepaint holes 1`] = `
.emotion-0 {
  color: red;
}

@media (min-width:920px) {
  .emotion-0 {
    color: blue;
  }
}

@media (min-width:1120px) {
  .emotion-0 {
    color: darkorchid;
  }
}

<div
  className="emotion-0"
>
  Basic
</div>
`;

exports[`facepaint holes 2`] = `
".css-196uzh {
  color: red;
}

@media (min-width:920px) {
  .css-196uzh {
    color: blue;
  }
}

@media (min-width:1120px) {
  .css-196uzh {
    color: darkorchid;
  }
}"
`;

exports[`facepaint more composition 1`] = `
.emotion-0 {
  margin-top: 1px;
  margin-top: 500px;
}

@media (min-width:420px) {
  .emotion-0 {
    margin-top: 2px;
  }
}

<div
  className="emotion-0"
/>
`;

exports[`facepaint more composition 2`] = `
".css-1lg6icu {
  margin-top: 1px;
}

@media (min-width:420px) {
  .css-1lg6icu {
    margin-top: 2px;
  }
}

.css-lk2z8m {
  margin-top: 500px;
}

.css-1874a9x {
  margin-top: 1px;
  margin-top: 500px;
}

@media (min-width:420px) {
  .css-1874a9x {
    margin-top: 2px;
  }
}"
`;

exports[`facepaint multiple 1`] = `
.emotion-0 {
  color: red;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 12px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media (min-width:420px) {
  .emotion-0 {
    color: green;
    display: block;
  }
}

@media (min-width:920px) {
  .emotion-0 {
    color: blue;
    display: inline-block;
  }
}

@media (min-width:1120px) {
  .emotion-0 {
    color: darkorchid;
    display: table;
  }
}

<div
  className="emotion-0"
>
  multiple
</div>
`;

exports[`facepaint multiple 2`] = `
".css-1lsvhjc {
  color: red;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 12px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media (min-width:420px) {
  .css-1lsvhjc {
    color: green;
    display: block;
  }
}

@media (min-width:920px) {
  .css-1lsvhjc {
    color: blue;
    display: inline-block;
  }
}

@media (min-width:1120px) {
  .css-1lsvhjc {
    color: darkorchid;
    display: table;
  }
}"
`;

exports[`facepaint nested 1`] = `
.emotion-0 {
  background-color: hotpink;
  text-align: center;
  width: 25%;
}

@media (min-width:420px) {
  .emotion-0 {
    width: 50%;
  }
}

@media (min-width:920px) {
  .emotion-0 {
    width: 75%;
  }
}

@media (min-width:1120px) {
  .emotion-0 {
    width: 100%;
  }
}

.emotion-0 .foo {
  color: red;
}

@media (min-width:420px) {
  .emotion-0 .foo {
    color: green;
  }
}

@media (min-width:920px) {
  .emotion-0 .foo {
    color: blue;
  }
}

@media (min-width:1120px) {
  .emotion-0 .foo {
    color: darkorchid;
  }
}

.emotion-0 .foo img {
  height: 10px;
}

@media (min-width:420px) {
  .emotion-0 .foo img {
    height: 15px;
  }
}

@media (min-width:920px) {
  .emotion-0 .foo img {
    height: 20px;
  }
}

@media (min-width:1120px) {
  .emotion-0 .foo img {
    height: 25px;
  }
}

<div
  className="emotion-0"
>
  <div
    className="foo"
  >
    foo
  </div>
  function
</div>
`;

exports[`facepaint nested 2`] = `
".css-rbuh8g {
  background-color: hotpink;
  text-align: center;
  width: 25%;
}

@media (min-width:420px) {
  .css-rbuh8g {
    width: 50%;
  }
}

@media (min-width:920px) {
  .css-rbuh8g {
    width: 75%;
  }
}

@media (min-width:1120px) {
  .css-rbuh8g {
    width: 100%;
  }
}

.css-rbuh8g .foo {
  color: red;
}

@media (min-width:420px) {
  .css-rbuh8g .foo {
    color: green;
  }
}

@media (min-width:920px) {
  .css-rbuh8g .foo {
    color: blue;
  }
}

@media (min-width:1120px) {
  .css-rbuh8g .foo {
    color: darkorchid;
  }
}

.css-rbuh8g .foo img {
  height: 10px;
}

@media (min-width:420px) {
  .css-rbuh8g .foo img {
    height: 15px;
  }
}

@media (min-width:920px) {
  .css-rbuh8g .foo img {
    height: 20px;
  }
}

@media (min-width:1120px) {
  .css-rbuh8g .foo img {
    height: 25px;
  }
}"
`;

exports[`facepaint nested arrays 1`] = `
.emotion-0 {
  color: red;
}

@media (min-width:420px) {
  .emotion-0 {
    color: blue;
  }
}

@media (min-width:920px) {
  .emotion-0 {
    color: darkorchid;
  }
}

<div
  className="emotion-0"
>
  Basic
</div>
`;

exports[`facepaint nested arrays 2`] = `
".css-1faqh3h {
  color: red;
}

@media (min-width:420px) {
  .css-1faqh3h {
    color: blue;
  }
}

@media (min-width:920px) {
  .css-1faqh3h {
    color: darkorchid;
  }
}"
`;

exports[`facepaint pseudo 1`] = `
.emotion-0 {
  background-color: hotpink;
  text-align: center;
  width: 25%;
}

.emotion-0:hover {
  width: 50%;
}

.emotion-0:active {
  width: 75%;
}

.emotion-0:focus {
  width: 100%;
}

.emotion-0 .foo {
  color: red;
}

.emotion-0 .foo:hover {
  color: green;
}

.emotion-0 .foo:active {
  color: blue;
}

.emotion-0 .foo:focus {
  color: darkorchid;
}

.emotion-0 .foo img {
  height: 10px;
}

.emotion-0 .foo img:hover {
  height: 15px;
}

.emotion-0 .foo img:active {
  height: 20px;
}

.emotion-0 .foo img:focus {
  height: 25px;
}

<div
  className="emotion-0"
>
  <div
    className="foo"
  >
    foo
  </div>
  function
</div>
`;

exports[`facepaint pseudo 2`] = `
".css-1guvnfu {
  background-color: hotpink;
  text-align: center;
  width: 25%;
}

.css-1guvnfu:hover {
  width: 50%;
}

.css-1guvnfu:active {
  width: 75%;
}

.css-1guvnfu:focus {
  width: 100%;
}

.css-1guvnfu .foo {
  color: red;
}

.css-1guvnfu .foo:hover {
  color: green;
}

.css-1guvnfu .foo:active {
  color: blue;
}

.css-1guvnfu .foo:focus {
  color: darkorchid;
}

.css-1guvnfu .foo img {
  height: 10px;
}

.css-1guvnfu .foo img:hover {
  height: 15px;
}

.css-1guvnfu .foo img:active {
  height: 20px;
}

.css-1guvnfu .foo img:focus {
  height: 25px;
}"
`;

exports[`facepaint repeating 1`] = `
.emotion-0 {
  color: red;
}

@media (min-width:420px) {
  .emotion-0 {
    color: blue;
  }
}

@media (min-width:11200px) {
  .emotion-0 {
    color: darkorchid;
  }
}

<div
  className="emotion-0"
>
  Basic
</div>
`;

exports[`facepaint repeating 2`] = `
".css-nhhz1q {
  color: red;
}

@media (min-width:420px) {
  .css-nhhz1q {
    color: blue;
  }
}

@media (min-width:11200px) {
  .css-nhhz1q {
    color: darkorchid;
  }
}"
`;

exports[`facepaint undefined 1`] = `
.emotion-0 {
  color: red;
}

@media (min-width:920px) {
  .emotion-0 {
    color: blue;
  }
}

@media (min-width:1120px) {
  .emotion-0 {
    color: darkorchid;
  }
}

<div
  className="emotion-0"
>
  Basic
</div>
`;

exports[`facepaint undefined 2`] = `
".css-196uzh {
  color: red;
}

@media (min-width:920px) {
  .css-196uzh {
    color: blue;
  }
}

@media (min-width:1120px) {
  .css-196uzh {
    color: darkorchid;
  }
}"
`;
